worker/compile/compile.go

```diff
@@ -0,0 +1,152 @@
+package compile
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/natsukagami/kjudge/models"
+	"github.com/pkg/errors"
+)
+
+// Compile performs compilation.
+// Returns whether the compilation succeeds.
+func Compile(c *CompileContext) (bool, error) {
+	action, err := getAction(c)
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	log.Printf("[WORKER] Compiling submission %v\n", c.Sub.ID)
+
+	// Now, create a temporary directory.
+	dir, err := os.MkdirTemp("", "*")
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+	defer action.Cleanup(dir)
+
+	// Prepare source and files
+	action.Source.Content = c.Sub.Source
+	action.Files = c.Files
+	if err := action.Prepare(dir); err != nil {
+		return false, err
+	}
+
+	// Perform compilation
+	result, messages := action.Perform(dir)
+	c.Sub.CompilerOutput = messages
+
+	if result {
+		// Success!
+		output, err := os.ReadFile(filepath.Join(dir, action.Output))
+		if err != nil {
+			return false, errors.WithStack(err)
+		}
+		c.Sub.CompiledSource = output
+	} else {
+		c.Sub.CompiledSource = nil
+		c.Sub.Verdict = models.VerdictCompileError
+	}
+	log.Printf("[WORKER] Compiling submission %v succeeded (result = %v).", c.Sub.ID, result)
+
+	return result, c.Sub.Write(c.DB)
+}
+
+// Try creating a compile action using available schemes
+func getAction(c *CompileContext) (*CompileAction, error) {
+	if action, err := CompileBatch(c); action != nil || err != nil {
+		return action, err
+	}
+	if action, err := CompileSingle(c); action != nil || err != nil {
+		return action, err
+	}
+	return nil, errors.Errorf("no scheme accepted context")
+}
+
+var recognizedFilenames = []string{
+	"statements.pdf", "statements.md", "compare", ".stages",
```

Comment:
I think you should also check for `compare.*` files, since you can upload one of them and let `kjudge` compile them for you ;)

Reply:
I think this should also be added to the notes on frontend.